src/map.rs

```diff
@@ -306,76 +327,142 @@ impl TileMap {
     /// tile_map.new_chunk(1);
     /// tile_map.new_chunk(2);
     /// ```
-    pub fn new_chunk<I: ToIndex>(&mut self, v: I) -> DimensionResult<()> {
+    pub fn new_chunk<I: ToIndex>(&mut self, v: I) -> MapResult<()> {
         let index = v.to_index(self.dimensions.width(), self.dimensions.height());
         self.dimensions.check_index(index)?;
 
-        let tiles = vec![
-            Tile::new(0);
-            (self.chunk_dimensions().width() * self.chunk_dimensions().height())
-                as usize
-        ];
-        self.events.send(MapEvent::Created { index, tiles });
+        let handle: Handle<Chunk> = Handle::weak(HandleId::random::<Chunk>());
+        self.chunks.insert(index, handle.clone_weak());
+
+        self.events.send(MapEvent::CreatedChunk { index, handle });
 
         Ok(())
     }
 
-    /// Constructs a new `Chunk` and stores it at a coordinate position with
-    /// tiles.
+    /// Adds a layer to the `TileMap` and inner chunks.
     ///
-    /// It requires that you give it either an index or a Vec2 or Vec3
-    /// coordinate as well as a vector of `Tile`s. It then automatically sets
-    /// both a sized mesh and chunk for use based on the parameters set in the
-    /// parent `TileMap`.
+    /// This method takes in a `LayerKind`, coordinate as either an index or
+    /// Vec2, as well as a specified Z layer that the layer needs to be set to.
+    /// If the layer is already the specified layer's kind, then nothing
+    /// happens.
+    ///
+    /// # Errors
+    ///
+    /// If a layer is set and a different layer already exists at that Z layer
+    /// then an error is returned regarding that.
     ///
     /// # Panics
     ///
-    /// This method will panic if you attempt to add a chunk to an out of bounds
-    /// index location or coordinate.
+    /// If the tile length is not the same length as the expected tile length
+    /// per chunk, then a panic is thrown. This is important as it will lead to
+    /// future panics if it is not caught here.
+    pub fn add_layer(&mut self, kind: LayerKind, z_layer: usize) -> MapResult<()> {
```

Comment:
A nice think about not having two different kinds of layers is that this function no longer needs to exist in the public API. You can just implicitly create the layer when a tile is added to it because no other parameter is required.

Reply:
Resolved with `add_layer` setting a default layerkind.